File: solutions/elixir/raindrops/1/lib/raindrops.ex
defmodule Raindrops do
  @sound_by_factor %{
    3 => "Pling",
    5 => "Plang",
    7 => "Plong"
  }
  @no_sound ""
  @is_factor true
  @not_factor false

  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    sounds = convert(number, @sound_by_factor)

    if sounds !== "",
      do: sounds,
      else: to_string(number)
  end

  defp convert(number, %{} = sound_by_factor),
    do:
      sound_by_factor
      |> Enum.map(&convert(number, &1))
      |> Enum.reject(&(&1 === @no_sound))
      |> Enum.join("")

  defp convert(number, {value, sound}),
    do: make_sound(is_factor?(number, value), sound)

  defp make_sound(@is_factor, sound),
    do: sound

  defp make_sound(@not_factor, sound),
    do: @no_sound

  defp is_factor?(number, value),
    do: rem(number, value) === 0
end

File: solutions/scala/raindrops/1/src/main/scala/Raindrops.scala
object Raindrops {
  val soundByFactors: Map[Int, String] = Map[Int, String](
    3 → "Pling",
    5 → "Plang",
    7 → "Plong"
  )

  val noSound = ""
  val isFactor = true
  val isNotFactor = false

  def convert(number: Int): String = {
    val sounds: String = soundByFactors.map(this.toSound(number, _))
      .filterNot(_ == noSound)
      .mkString("")

    if (!sounds.isBlank) {
      sounds
    }
    else {
      number.toString
    }
  }

  private def toSound(number: Int, soundKeyValue: (Int, String)): String = {
    val (key, sound) = soundKeyValue
    if (isFactor(number, key)) {
      sound
    }
    else {
      noSound
    }
  }

  private def isFactor(number: Int, value: Int) =
    number % value == 0
}

